app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -4,18 +4,23 @@ class OmniauthCallbacksController < ActionController::Base
 
   skip_before_action :verify_authenticity_token, only: :callback
 
+  before_action :verify_callback_contents, only: :callback
+
   def login
     render :login
   end
 
   def callback
-    case params[:provider]
-    when 'google_oauth2'
-      google_callback
-    else
-      render text: 'Unknown oauth provider', status: :forbidden
-      return
-    end
+    set_user_photo
+    set_user_metadata
+
+    user.save!
+
+    session[:user_id] = user.id
+
+    redirect_to params[:return_to] || :portal
```

Comment:
we should do some validations on the return_to param. At least make sure it's only a relative URL otherwise this can be used as an attack vector for someone to get people onto a malicious website.